File: templates/observability/sentry/python/sentry_init.py
"""octc-platform — Sentry init template for Python services.

Política: docs/observability/OBSERVABILITY.md
"""
from __future__ import annotations

import os
from typing import Any

import sentry_sdk


_SENSITIVE_KEYS = (
    "email",
    "phone",
    "name",
    "address",
    "id_document",
    "ssn",
    "credit_card",
    "auth",
    "authorization",
    "cookie",
    "session",
    "password",
)


def _scrub(value: Any) -> Any:
    if isinstance(value, dict):
        return {
            k: ("[redacted]" if any(s in k.lower() for s in _SENSITIVE_KEYS) else _scrub(v))
            for k, v in value.items()
        }
    if isinstance(value, list):
        return [_scrub(v) for v in value]
    return value


def _before_send(event: dict[str, Any], _hint: Any) -> dict[str, Any]:
    if "request" in event:
        event["request"] = _scrub(event["request"])
    if "extra" in event:
        event["extra"] = _scrub(event["extra"])
    if "contexts" in event:
        event["contexts"] = _scrub(event["contexts"])
    return event


def init_octc_sentry() -> None:
    sentry_sdk.init(
        dsn=os.environ.get("SENTRY_DSN"),
        environment=os.environ.get("SENTRY_ENVIRONMENT", "development"),
        release=os.environ.get("SENTRY_RELEASE"),
        traces_sample_rate=float(os.environ.get("SENTRY_TRACES_SAMPLE_RATE", "0.2")),
        profiles_sample_rate=float(os.environ.get("SENTRY_PROFILES_SAMPLE_RATE", "0.05")),
        send_default_pii=False,
        before_send=_before_send,
    )

File: templates/observability/sentry/flutter/sentry_init.dart
// octc-platform — Flutter Sentry init template
// Política: docs/observability/OBSERVABILITY.md

import 'package:flutter/foundation.dart';
import 'package:sentry_flutter/sentry_flutter.dart';

const _sensitiveKeys = <String>{
  'email',
  'phone',
  'name',
  'address',
  'id_document',
  'ssn',
  'credit_card',
  'auth',
  'authorization',
  'cookie',
  'session',
  'password',
};

Object? _scrub(Object? value) {
  if (value is Map) {
    return value.map((k, v) {
      final key = k.toString().toLowerCase();
      if (_sensitiveKeys.any((s) => key.contains(s))) {
        return MapEntry(k, '[redacted]');
      }
      return MapEntry(k, _scrub(v));
    });
  }
  if (value is List) {
    return value.map(_scrub).toList();
  }
  return value;
}

Future<void> initOctcSentry(Future<void> Function() runApp) async {
  await SentryFlutter.init(
    (options) {
      options.dsn = const String.fromEnvironment('SENTRY_DSN');
      options.environment = const String.fromEnvironment(
        'SENTRY_ENVIRONMENT',
        defaultValue: 'development',
      );
      options.release = const String.fromEnvironment('SENTRY_RELEASE');
      options.tracesSampleRate = 0.1;
      options.profilesSampleRate = 0.05;
      options.sendDefaultPii = false;
      options.beforeSend = (event, hint) async {
        return event.copyWith(
          extra: _scrub(event.extra) as Map<String, dynamic>?,
          contexts: event.contexts,
        );
      };
    },
    appRunner: runApp,
  );
}
